### PR TITLE
[SMALLFIX] Fix a compatibility issue for journal replay with security mode change.

### PR DESCRIPTION
**Pull request by ifcharming**
Tested locally: restarting a cluster by enabling security from `NOSASL` to `SIMPLE`, without formatting the journal.

Without this change, the journal generated with security off would not be replayed correctly into a new version with security on.

### CODE REVIEW

**Comment by ifcharming** [comment]
@gpang @apc999 @calvinjia

**Comment by ifcharming** [comment]
@apc999  yes, that will be in a separate PR.

File: core/server/src/main/java/alluxio/master/file/meta/InodeTree.java [hunk 1]
```diff
@@ -831,15 +831,24 @@ public void addInodeFromJournal(JournalEntry entry) throws AccessControlExceptio
...
           // user is not the owner of journal root entry
           throw new AccessControlException(
               ExceptionMessage.PERMISSION_DENIED.getMessage("Unauthorized user on root"));
         }
         mInodes.clear();
         mPinnedInodeFileIds.clear();
         mRoot = directory;
+        // If journal entry has no security enabled, change the replayed inode permission to be 0777
+        // for backwards-compatibility.
+        if (SecurityUtils.isSecurityEnabled() && mRoot != null && mRoot.getOwner().isEmpty()
```
Comment (gpang) [inline core/server/src/main/java/alluxio/master/file/meta/InodeTree.java]: NIT: is it possible for `mRoot` (or `directory`) to be `null`?

File: core/server/src/main/java/alluxio/master/file/meta/InodeTree.java [hunk 2]
```diff
@@ -864,6 +873,12 @@ private void addInodeFromJournalInternal(Inode<?> inode) {
     }
     parentDirectory.addChild(inode);
     mInodes.add(inode);
+    // If journal entry has no security enabled, change the replayed inode permission to be 0777
+    // for backwards-compatibility.
+    if (SecurityUtils.isSecurityEnabled() && inode != null && inode.getOwner().isEmpty()
```
Comment (gpang) [inline core/server/src/main/java/alluxio/master/file/meta/InodeTree.java]: NIT: is it possible for `inode` to be `null`?

File: core/server/src/main/java/alluxio/master/file/meta/InodeTree.java (hunk 1 above)
Comment (ifcharming) [inline core/server/src/main/java/alluxio/master/file/meta/InodeTree.java]: AFAICT, it's possible for `mRoot` to be `null`:
`InodeTree#initializeRoot` is only called by `FileSystemMaster` during `start()`, while `addInodeFromJournal` (and `processJournalCheckpoint`) can be called from any `AbstractMaster` via `JournalTailer`.

File: core/server/src/main/java/alluxio/master/file/meta/InodeTree.java (hunk 2 above)
Comment (ifcharming) [inline core/server/src/main/java/alluxio/master/file/meta/InodeTree.java]: This one perhaps not possible, unless `InodeFileEntry` can represent a null `InodeFile`

File: core/server/src/main/java/alluxio/master/file/meta/InodeTree.java (hunk 1 above)
Comment (gpang) [inline core/server/src/main/java/alluxio/master/file/meta/InodeTree.java]: yes, but on the previous line, `mRoot` is set to `directory`.

**Comment by ifcharming** [comment]
@gpang I've replied to your questions, but not totally sure. Maybe can we leave the `not-null` check to be safe?

**Comment by gpang** [comment]
Ok, we can leave it in. I just want to understand when objects could be `null`.

File: core/server/src/main/java/alluxio/master/file/meta/InodeTree.java (hunk 1 above)
Comment (ifcharming) [inline core/server/src/main/java/alluxio/master/file/meta/InodeTree.java]: Oh right, I was referring to the first check on line 837. Yeah, for those two after setting `mRoot`, it depends on whether it's allowed to have a journal entry referring to `null`.

**Pull request merged**